diff --git a/src/Form/FieldValidationRuleAddForm.php b/src/Form/FieldValidationRuleAddForm.php
index 7eb45cb..bbd55c1 100644
--- a/src/Form/FieldValidationRuleAddForm.php
+++ b/src/Form/FieldValidationRuleAddForm.php
@@ -46,7 +46,7 @@ class FieldValidationRuleAddForm extends FieldValidationRuleFormBase {
   /**
    * {@inheritdoc}
    */
-  public function buildForm(array $form, FormStateInterface $form_state, FieldValidationRuleSetInterface $field_validation_rule_set = NULL, $field_validation_rule = NULL) {
+  public function buildForm(array $form, FormStateInterface $form_state, FieldValidationRuleSetInterface $field_validation_rule_set = NULL, $field_validation_rule = NULL, $field_name='') {
     $form = parent::buildForm($form, $form_state, $field_validation_rule_set, $field_validation_rule);
     //drupal_set_message('term_id:' . var_export($field_validation_rule));
     $form['#title'] = $this->t('Add %label field validation rule', array('%label' => $this->fieldValidationRule->label()));
diff --git a/src/Form/FieldValidationRuleEditForm.php b/src/Form/FieldValidationRuleEditForm.php
index 68edb71..f1cab4b 100644
--- a/src/Form/FieldValidationRuleEditForm.php
+++ b/src/Form/FieldValidationRuleEditForm.php
@@ -18,7 +18,7 @@ class FieldValidationRuleEditForm extends FieldValidationRuleFormBase {
   /**
    * {@inheritdoc}
    */
-  public function buildForm(array $form, FormStateInterface $form_state, FieldValidationRuleSetInterface $field_validation_rule_set = NULL, $field_validation_rule = NULL) {
+  public function buildForm(array $form, FormStateInterface $form_state, FieldValidationRuleSetInterface $field_validation_rule_set = NULL, $field_validation_rule = NULL, $field_name='') {
     $form = parent::buildForm($form, $form_state, $field_validation_rule_set, $field_validation_rule);
 
     $form['#title'] = $this->t('Edit %label rule', array('%label' => $this->fieldValidationRule->label()));
